pkg/controller/controller.go

```diff
@@ -591,7 +610,61 @@ func removePrefixedParameters(param map[string]string) (map[string]string, error
 	return newParam, nil
 }
 
+// getVolumeContentSource is a helper function to process provisioning requests that include a DataSource
+// currently we provide Snapshot and PVC, the default case allows the provisioner to still create a volume
+// so that an external controller can act upon it.   Additional DataSource types can be added here with
+// an appropriate implementation function
 func (p *csiProvisioner) getVolumeContentSource(options controller.VolumeOptions) (*csi.VolumeContentSource, error) {
+	switch options.PVC.Spec.DataSource.Kind {
+	case snapshotKind:
+		return p.getSnapshotSource(options)
+	case pvcKind:
+		return p.getPVCSource(options)
+	default:
+		// For now we shouldn't pass other things to this function, but treat it as a noop and extend as needed
+		return nil, nil
+	}
+}
+
+// getPVCSource verifies DataSource.Kind of type PersistentVolumeClaim, making sure that the requested PVC is available/ready
+// returns the VolumeContentSource for the requested PVC
+func (p *csiProvisioner) getPVCSource(options controller.VolumeOptions) (*csi.VolumeContentSource, error) {
+	sourcePVC, err := p.client.CoreV1().PersistentVolumeClaims(options.PVC.Namespace).Get(options.PVC.Spec.DataSource.Name, metav1.GetOptions{})
+	if err != nil {
+		return nil, fmt.Errorf("error getting PVC %s from api server: %v", options.PVC.Spec.DataSource.Name, err)
+	}
+	if string(sourcePVC.Status.Phase) != "Bound" {
+		return nil, fmt.Errorf("the PVC DataSource %s must have a status of Bound.  Got %v", options.PVC.Spec.DataSource.Name, sourcePVC.Status)
+	}
+	if sourcePVC.ObjectMeta.DeletionTimestamp != nil {
+		return nil, fmt.Errorf("the PVC DataSource %s is currently being deleted", options.PVC.Spec.DataSource.Name)
+	}
+	if sourcePVC.Spec.StorageClassName != options.PVC.Spec.StorageClassName {
+		return nil, fmt.Errorf("the source PVC and destination PVCs must be in the same storage class for cloning.  Source is in %v, but new PVC is in %v",
+			sourcePVC.Spec.StorageClassName, options.PVC.Spec.StorageClassName)
+	}
+	capacity := options.PVC.Spec.Resources.Requests[v1.ResourceName(v1.ResourceStorage)]
+	requestedSize := capacity.Value()
+	if requestedSize < int64(sourcePVC.Spec.Size()) {
```

Comment:
@j-griffith  There could be scenarios where `resizer` is processing an expansion request in the backend. So, may be checking current PVC's size and capacity could help as well.

Reply:
Sorry, i'm not sure I follow; what do you mean the "current PVC's size and capacity"?  The destination has yet to be created at this stage, so the only object that's relevant is the sourcePVC and the requested size; do you mean a resize in progress on the specified source volume?

Reply:
>do you mean a resize in progress on the specified source volume?
Exactly, the resizer took the request and working with the driver to get it into desired state.

Reply:
Shouldn't checking `sourcePVC.Spec` alone be sufficient since it will be the largest possible value (it will be the desired size)? Validation prevents setting that value to a smaller value right?

Reply:
> @j-griffith There could be scenarios where `resizer` is processing an expansion request in the backend. So, may be checking current PVC's size and capacity could help as well.

@humblec so I've thought about this a bit; and yes there's certainly a potential for a race here.  That being said, we have that condition across other operations already (snapshots for example) and the challenge that's been discussed is coordinating any sort of locking/checks across the different control planes.  Not to mention as @jsafrane pointed out, even across CSI provisioner nodes.

IMO this is a larger problem that needs solved independent of the cloning implementation.  Whether that's new phases on the PVC, some other even tracking, or (as I think we've done thus far) the "storage device plugins problem".  It sounds like this is something that's been brought up in other context and is being discussed, but we have some long term work to do here I think.